modules/quantcastIdSystem.js

```diff
@@ -7,11 +7,156 @@
+    } else if (!fpa) {
+      var expires = new Date(now.getTime() + (cookieExpTime * 86400000)).toGMTString();
+      fpa = 'B0-' + Math.round(Math.random() * 2147483647) + '-' + et;
+      fpan = '1';
+      storage.setCookie(QUANTCAST_FPA, fpa, expires, '/', domain, null);
+    }
+
+    firstPartyParamStrings = `&fpan=${fpan}&fpa=${fpa}`;
+    gdprParamStrings = '&gdpr=0';
+    if (gdprPrivacyString && typeof gdprPrivacyString.gdprApplies === 'boolean' && gdprPrivacyString.gdprApplies) {
+      gdprParamStrings = `gdpr=1&gdpr_consent=${gdprPrivacyString.consentString}`;
+    }
+    if (usPrivacyString && typeof usPrivacyString === 'string') {
+      usPrivacyParamString = `&us_privacy=${usPrivacyString}`;
+    }
+
+    let url = DOMAIN_QSERVE +
+    '?d=' + domain +
+    '&et=' + et +
+    '&tzo=' + tzo +
+    '&client_id=' + clientId +
+    '&a=' + PREBID_PCODE +
+    usPrivacyParamString +
+    gdprParamStrings +
+    firstPartyParamStrings;
+
+    triggerPixel(url);
+  }
+};
+
+export function hasGDPRConsent(gdprConsent) {
+  // Check for GDPR consent for purpose 1 and 10, and drop request if consent has not been given
+  // Remaining consent checks are performed server-side.
+  if (gdprConsent && typeof gdprConsent.gdprApplies === 'boolean' && gdprConsent.gdprApplies) {
+    if (!gdprConsent.vendorData) {
+      return false;
+    }
+    if (gdprConsent.apiVersion === 1) {
+      return checkTCFv1(gdprConsent.vendorData);
+    }
+    if (gdprConsent.apiVersion === 2) {
+      return checkTCFv2(gdprConsent.vendorData);
+    }
+  }
+  return true;
+}
+
+export function checkTCFv1(vendorData) {
+  var vendorConsent = vendorData.vendorConsents && vendorData.vendorConsents[QUANTCAST_VENDOR_ID];
+  var purposeConsent = vendorData.purposeConsents && vendorData.purposeConsents[PURPOSE_DATA_COLLECT];
+
+  return !!(vendorConsent && purposeConsent);
+}
+
+export function checkTCFv2(vendorData) {
+  var vendorConsent = vendorData.vendor && vendorData.vendor.consents && vendorData.vendor.consents[QUANTCAST_VENDOR_ID];
+  var vendorInterest = vendorData.vendor.legitimateInterests && vendorData.vendor.legitimateInterests[QUANTCAST_VENDOR_ID];
+  var restrictions = vendorData.publisher ? vendorData.publisher.restrictions : {};
+
+  // Restrictions for purpose 1
+  var qcRestriction = restrictions && restrictions[PURPOSE_DATA_COLLECT]
+    ? restrictions[PURPOSE_DATA_COLLECT][QUANTCAST_VENDOR_ID]
+    : null;
+
+  var purposeConsent = vendorData.purpose && vendorData.purpose.consents && vendorData.purpose.consents[PURPOSE_DATA_COLLECT];
+
+  // No consent, not allowed by publisher or requires legitimate interest
+  if (!vendorConsent || !purposeConsent || qcRestriction === 0 || qcRestriction === 2) {
+    return false;
+  }
+
+  // Restrictions for purpose 10
+  qcRestriction = restrictions && restrictions[PURPOSE_PRODUCT_IMPROVEMENT]
+    ? restrictions[PURPOSE_PRODUCT_IMPROVEMENT][QUANTCAST_VENDOR_ID]
+    : null;
+
```

Comment:
need logic for the case where there are no restrictions

Reply:
(just check Quantcast has legitimate interest for purpose 10 `PURPOSE_PRODUCT_IMPROVEMENT`)